cirq/contrib/acquaintance/permutation.py

```diff
@@ -224,14 +227,30 @@ def update_mapping(mapping: Dict[ops.Qid, LogicalIndex],
 
 def get_logical_operations(operations: 'cirq.OP_TREE',
                            initial_mapping: Dict[ops.Qid, ops.Qid]
-                          ) -> Iterable['cirq.Operation']:
+                          ) -> Optional[List['cirq.Operation']]:
+    """Gets the logical operations specified by the physical operations and
+    initial mapping.
+
+    If a non-permutation physical operation acts on an unmapped qubit, returns
+    None.
+
+    Args:
+        operations: The physical operations.
+        initial_mapping: The initial mapping of physical to logical qubits.
+    """
     mapping = initial_mapping.copy()
+    logical_ops = []
     for op in cast(Iterable['cirq.Operation'], ops.flatten_op_tree(operations)):
         if (isinstance(op, ops.GateOperation) and
                 isinstance(op.gate, PermutationGate)):
             op.gate.update_mapping(mapping, op.qubits)
         else:
-            yield op.transform_qubits(mapping.__getitem__)
+            for q in op.qubits:
+                if mapping.get(q) is None:
+                    return None
```

Comment:
I feel like this should raise an error instead of returning None.

Reply:
That was my original impulse as well, but this allows `is_valid_routing` to handle it without catching exceptions.

Reply:
I could add a "strict" flag that throw an error in such cases.

I also thought about returning a list the same length as `operations` but with `None` where there is a logical operation on an unmapped qubit, but I didn't see a use case for such precision.

Reply:
I would raise an error and have `is_valid_routing` handle an exception.

Reply:
Why? `None` indicates that there isn't a well-defined answer. It's also not iterable, so a caller that doesn't account for that return value would get an error anyway.

Reply:
I think that returning None is surprising behavior. I think that in most cases a user calling this function expects to get something as the output and it's better to give a helpful error message if they got the input wrong.

Reply:
Done